pkg/kv/kvserver/rafttermcache/term_cache.go

```diff
@@ -0,0 +1,384 @@
+	// the entry at lastIndex has the same term as TermCache.cache's
+	// last entry's term.
+	//
+	// lastIndex is set to 0 when TermCache is created.
+	//
+	// lastIndex is first updated with the storage engine's
+	// term: r.shMu.lastTermNotDurable and index: r.shMu.lastIndexNotDurable
+	// as the first entry. This entry is the last entry in the raftLog's
+	// persisted part.
+	//
+	// in the special case when raftLog is completely empty,
+	// (can happen when deploying the database for the first time with no data)
+	// the TruncatedState == LastEntryID
+	// (the TruncatedState represents the highest applied-to-state machine entry)
+	//
+	//
+	// lastIndex is updated when entries are appended to the TermCache.
+	//
+	// Invariants (except for on the creation of the TermCache):
+	//
+	// For callers, lastIndex is always populated (lastIndex >= 0),
+	// and there is always at least one entry being covered in the termCache.
+	//
+	// lastIndex mirrors the last index in the storage part of raftLog.
+	lastIndex kvpb.RaftIndex
+	// mu is used to protect the term cache from concurrent accesses.
+	// Which can happen for example when the following is called concurrently:
+	// - logstore.LoadTerm() in replicaLogStorage.termLocked(),
+	// - ScanAppend() in logstore/storeEntriesAndCommitBatch().
+	// TODO(hakuuww): using a RWMutex is one way, and can guarantee safeness.
+	//  But maybe we can just lock replica.mu externally when calling ScanAppend()
+	//  and avoid using locks when calling LoadTerm().
+	mu syncutil.RWMutex
+}
+
+// errInvalidEntryID is returned when the supplied entryID
+// is invalid for the operation. Indicates inconsistent log entries.
+var errInvalidEntryID = errors.New("invalid entry ID")
+
+// NewTermCache initializes a TermCache with a fixed maxSize.
+func NewTermCache(size uint64) *TermCache {
+	return &TermCache{
+		cache:     make([]entryID, 0, size),
+		maxSize:   size,
+		lastIndex: 0,
+	}
+}
+
+// Term returns the entry term based on the given entry index.
+// Returns error if not in the term cache.
+func (tc *TermCache) Term(
+	ctx context.Context, index kvpb.RaftIndex,
+) (term kvpb.RaftTerm, found bool) {
+	tc.mu.RLock()
+	defer tc.mu.RUnlock()
+
+	if len(tc.cache) == 0 {
+		log.Fatalf(ctx, "term cache is empty when calling Term(), should not happen")
+		return 0, false
+	}
+
+	if index < tc.firstEntry().index ||
+		index > tc.lastIndex {
+		return 0, false
+	}
+
+	// in last term of term cache, index <= tc.lastIndex
+	if index >= tc.lastEntry().index {
+		return tc.lastEntry().term, true
+	}
+
+	for i := len(tc.cache) - 2; i >= 0; i-- {
+		if index >= tc.cache[i].index {
+			return tc.cache[i].term, true
+		}
+	}
+
+	return 0, false
+}
+
+// ClearTo clears entries from the TermCache with index strictly less than hi.
+// If hi is above the lastIndex, the whole term cache is cleared.
+// Mirrors the clearTo function in raftentry/cache.go
+func (tc *TermCache) ClearTo(hi kvpb.RaftIndex) {
```

Comment:
Pretty sure this function can be squashed significantly, and we don't need any special cases. E.g.:

```golang

if index <= tc.cache[0].index {
	return // nothing to do
}
slot := 1
for ln := len(tc.cache); slot < ln; slot++ {
	if index < tc.cache[slot].index {
		break
	}
}
tc.cache = tc.cache[slot-1:]
tc.cache[0].index = index
```

This adds to my earlier point that making invariants is helpful. In this case, the `len(tc.cache) != 0` invariant.